wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/CheEnvironmentEngine.java

```diff
@@ -0,0 +1,954 @@
+                                            machineId,
+                                            environmentHolderCopy.name,
+                                            creator,
+                                            false,
+                                            environmentHolderCopy.logger);
+
+            replaceMachine(instance);
+
+            eventService.publish(newDto(MachineStatusEvent.class)
+                                         .withEventType(MachineStatusEvent.EventType.RUNNING)
+                                         .withDev(machineConfig.isDev())
+                                         .withMachineName(machineConfig.getName())
+                                         .withMachineId(instance.getId())
+                                         .withWorkspaceId(workspaceId));
+
+            return instance;
+        } catch (Exception e) {
+            removeMachine(workspaceId, machineId);
+
+            if (instance != null) {
+                try {
+                    instance.destroy();
+                } catch (Exception ignore) {
+
+                }
+            }
+
+            eventService.publish(newDto(MachineStatusEvent.class)
+                                         .withEventType(MachineStatusEvent.EventType.ERROR)
+                                         .withDev(machineConfig.isDev())
+                                         .withMachineName(machineConfig.getName())
+                                         .withMachineId(machineId)
+                                         .withWorkspaceId(workspaceId));
+
+            throw e;
+        }
+    }
+
+    /**
+     * Stops machine in running environment.
+     *
+     * @param workspaceId
+     *         ID of workspace of environment that owns machine
+     * @param machineId
+     *         ID of machine that should be stopped
+     * @throws NotFoundException
+     *         if machine in not found in environment
+     * @throws ConflictException
+     *         if environment is not running
+     * @throws ConflictException
+     *         if stop of dev machine is requested
+     * @throws ServerException
+     *         if other error occurs
+     */
+    public void stopMachine(String workspaceId, String machineId) throws NotFoundException,
+                                                                         ServerException,
+                                                                         ConflictException {
+        Instance targetMachine = null;
+        try (StripedLocks.WriteLock lock = stripedLocks.acquireWriteLock(workspaceId)) {
+            EnvironmentHolder environmentHolder = environments.get(workspaceId);
+            if (environmentHolder == null || environmentHolder.status != EnvStatus.RUNNING) {
+                throw new ConflictException(format("Environment '%s' is not running", workspaceId));
+            }
+            for (Instance machine : environmentHolder.machines) {
```

Comment:
It looks like you can simplify this code by reusing `removeFirstMatching` method